exec/eval.go

```diff
@@ -121,21 +127,10 @@ func Eval(ctx context.Context, executor Executor, roots []*Task, group *status.G
 					err = task.Wait(ctx)
 				}
 				if runner {
-					// Only the runner bookkeeps consecutiveLost to avoid
-					// double-counting task loss.
-					switch task.state {
-					case TaskOk:
-						task.consecutiveLost = 0
-					case TaskLost:
-						task.consecutiveLost++
-						if task.consecutiveLost >= maxConsecutiveLost {
-							// We've lost this task too many times, so we
-							// consider it in error.
-							task.state = TaskErr
-							task.err = fmt.Errorf("lost on %d consecutive attempts", task.consecutiveLost)
-							task.Status.Printf(task.err.Error())
-							task.Broadcast()
-						}
+					if enableMaxConsecutiveLost {
+						// Only the runner bookkeeps consecutiveLost to avoid
+						// double-counting task loss.
+						updateConsecutiveLost(task)
```

Comment:
I thought it was a bit overwhelming, especially because this code is really low-density w.r.t. affecting the already-tricky-to-understand flow of this code. However, you're a better judge of the overwhelming-ness than me, as I've been staring at this code for quite some time. I will either restore it, or try a different refactoring.

Reply:
It might depend on reading style. I tend to read "defensively", assuming there's something going on in each helper and often jumping over, which I guess causes me to understand new code in a combination of top-down (starting from `Eval`) and bottom-up (reading each `helper` as they're called) way. Perhaps if I had the right one-ness with the codebase the helpers would be more obvious to me, but as it is now some of the helpers struck me as slightly confusingly named (`schedule` doesn't do anything with the executor, `add` has an argument that its documentation says nothing about, there are two `(d|D)one`s that are quite different in nature). For what it's worth, I think `updateConsecutiveLost` is a much clearer name.

I've convinced myself that `updateConsecutiveLost` is ok as-is.

Reply:
I think your first instinct is reasonable, and I don't think it's a particularly compelling refactor that need to be part of this PR. I'll restore it for now, and if we want to refactor it, let's do it in a separate pass so as to not overload this change. The same goes for the naming of the other things you mention.